app/src/main/java/sk/inloop/support/v4/app/FragmentStatePagerAdapter.java

```diff
@@ -0,0 +1,248 @@
+        return new ArrayList<Fragment>();
+    }
+
+    /**
+     * Return the Fragment associated with a specified position.
+     */
+    public abstract Fragment getItem(int position);
+
+    @Override
+    public void startUpdate(ViewGroup container) {
+        if (container.getId() == View.NO_ID) {
+            throw new IllegalStateException("ViewPager with adapter " + this + " requires a view id");
+        }
+    }
+
+    @Override
+    public Object instantiateItem(ViewGroup container, int position) {
+        Fragment result = tryGetExistingFragment(position); //performance, if fragment has been initialized before, just return it
+        if (result != null) {
+            return result;
+        }
+
+        Fragment fragment = getItem(position);
+        if (DEBUG) {
+            Log.v(TAG, "Adding item #" + position + ": f=" + fragment);
+        }
+        Fragment.SavedState fss = tryGetSavedState(position);
+        if (fss != null) {
+            fragment.setInitialSavedState(fss);
+        }
+        setupFragmentAsSecondary(fragment);
+        addFragmentToList(position, fragment);
+        getOrCreateFragmentTransaction().add(container.getId(), fragment);
+
+        return fragment;
+    }
+
+    /**
+     * Add given {@link Fragment} to fragments list. Ensure, that this fragment will be found on given 'position' anytime (except if fragment will be removed).
+     */
+    protected void addFragmentToList(int position, @Nullable Fragment fragment) {
+        while (mFragments.size() <= position) {
+            //TODO: could be more effective???
+            mFragments.add(null);
+        }
+        mFragments.set(position, fragment);
+    }
+
+    /**
+     * Get existing {@link FragmentTransaction}. If transaction is not initialized, create it.
+     */
+    @NonNull
+    protected FragmentTransaction getOrCreateFragmentTransaction() {
+        if (mCurTransaction == null) {
+            mCurTransaction = mFragmentManager.beginTransaction();
+        }
+        return mCurTransaction;
+    }
+
+    /**
+     * Return saved {@link android.support.v4.app.Fragment.SavedState} for given 'position'. If state is not saved, NULL is returned.
+     */
+    @Nullable
+    protected Fragment.SavedState tryGetSavedState(int position) {
+        if (position >= 0 && mSavedState.size() > position) {
+            return mSavedState.get(position);
+        }
+        return null;
+    }
+
+    /**
+     * Return existing {@link Fragment} at defined 'position'. If fragment does not exist, NULL is returned.
+     */
+    @Nullable
+    protected Fragment tryGetExistingFragment(int position) {
+        if (position >= 0 && mFragments.size() > position) {
+            return mFragments.get(position);
+        }
+        return null;
+    }
+
+    @Override
+    public void destroyItem(ViewGroup container, int position, Object object) {
+        Fragment fragment = (Fragment) object;
+        if (DEBUG) {
```

Comment:
sorry, I forgot this while refactoring (this is from original FragmentState.. impl). Is Log.isloggable() accepted?